code-studio/src/js/callouts.js

```diff
@@ -112,16 +124,49 @@ module.exports = function createCallouts(callouts) {
     }
 
     if (callout.on) {
-      $(window).on(callout.on, function () {
-        if (!callout.seen && $(selector).length > 0) {
+      $(window).on(callout.on, function (e, action) {
+        if (!config.codeStudio.selector) {
+          config.codeStudio.selector = selector;
+        }
+        var lastSelector = config.codeStudio.selector;
+        $(window).trigger('prepareforcallout', [config.codeStudio]);
+        if (lastSelector !== config.codeStudio.selector && $(lastSelector).length > 0) {
+          $(lastSelector).qtip(config).qtip('destroy');
+        }
+        if ($(config.codeStudio.selector).length > 0) {
+          if (action === 'hashchange' || action === 'hashinit' || !callout.seen) {
+            $(config.codeStudio.selector).qtip(config).qtip('show');
+          }
           callout.seen = true;
-          $(selector).qtip(config).qtip('show');
         }
       });
-    } else {
+    } else if (!callout.seen) {
       $(selector).qtip(config).qtip('show');
     }
   });
+
+  // Insert a hashchange handler to detect triggercallout= hashes and fire
+  // appropriate events to open the callout
+  function detectTriggerCalloutOnHash(event) {
```

Comment:
I considered the "transform markdown at runtime" approach, but thought it would be nice to support this URL semantic directly (because jumping within the same document feels sort of like what hash URLs were intended for). Given that the scenario was "make it possible for HTML on the page to contain a link to jump to something within the same page" it feels like a good fit. I wouldn't want to use this as the normal method for communication on the page :) But if we ever did need to do something similar in the future, we could centralize one place to listen for the hashchange event, parse what's after the hash, and delegate as appropriate.

Regarding tests, are you thinking a UI test? that would cover most of the code and theoretically verify some of the toggline and scrolling behavior.